### DISCUSSION TITLE
Adding a TimeSpan to a DateTime, can it be done, and is it the right thing to do?

### DISCUSSION THREAD

**Discussion by BenjaBobs** [linq2db/linq2db#4106] [Q&A] [answered]
## Question
Is there a way to add a `TimeSpan` to a `DateTime` in an update statement?
We're trying to update a _NextTimeToDoStuff_ column, with a configurable interval stored as a `TimeSpan` (`DataType.Int64` in the server).
If not, is it because it's not a smart thing to do?
Should we be doing something else instead?

## Context

Example class
```cs
[Table("StocklistDownloadInfos")]
public class StocklistDownloadInfo
{
  [Column(IsPrimaryKey = true)
  public int Id { get; set; }

  [Column(Length = 500)]
  public string Url { get; set; }

  [Column]
  public DateTime NextDownloadAt { get; set; }

  [Column]
  public TimeSpan DownloadInterval { get; set; }
}
```

With `TimeSpan` handling inspired by https://stackoverflow.com/a/41979971/4823475:
```cs
// Store TimeSpan as Int64 in the database, inspired by https://stackoverflow.com/a/41979971/4823475.
mappingSchema.SetDataType(typeof(TimeSpan), DataType.Int64);
mappingSchema.SetDataType(typeof(TimeSpan?), DataType.Int64);
```

What we're trying to do is basically adding the `DownloadInterval` to the `NextDownloadAt`.

```cs
await connection
    .GetTable<StocklistDownloadInfo>()
    .Where(x => x.Id == someId)
    .Set(x => x.NextDownloadAt, x => x.NextDownloadAt + x.DownloadInterval)
    .UpdateAsync();
```

This does not work with our current setup and yields the following error:
```
System.ArgumentOutOfRangeException: Valid values are between 0 and 999, inclusive. (Parameter 'millisecond')

System.ArgumentOutOfRangeException
Valid values are between 0 and 999, inclusive. (Parameter 'millisecond')
   at System.DateTime.ThrowMillisecondOutOfRange()
   at System.DateTime..ctor(Int32 year, Int32 month, Int32 day, Int32 hour, Int32 minute, Int32 second, Int32 millisecond)
   at Microsoft.Data.Sqlite.SqliteValueReader.FromJulianDate(Double julianDate)
   at Microsoft.Data.Sqlite.SqliteValueReader.GetDateTime(Int32 ordinal)
   at Microsoft.Data.Sqlite.SqliteDataReader.GetDateTime(Int32 ordinal)
   at lambda_method780(Closure, IQueryRunner, DbDataReader)
   at LinqToDB.Linq.QueryRunner.Mapper`1.Map(IDataContext context, IQueryRunner queryRunner, DbDataReader dataReader, ReaderMapperInfo& mapperInfo)
```

We're using an Sqlite database to run our tests, so that's where I've encountered it, I haven't tested on other server types.
I've noticed that there are some `Sql.DateAdd()` methods, but I don't see one that takes in a `TimeSpan` as parameter.

**Comment by MaceWindu** [comment, marked as answer]
I don't think we provide such operation out-of-box, but you can add own mapping to sql.

1. You can define custom function
 
```cs
[Sql.Expression("your_custom_sql_here")]
public static DateTime MyAddFunction(DateTime date, TimeSpan interval) => throw new InvalidOperationException("server-side only method");
```

2. Override binary operation mapping (but I would not recommend it as it will hide) using `Expressions.MapBinary`, but I would not recommend it as it will override old behavior and requires function from (1) anyways

**Comment by igor-tkachev** [comment]
```cs
[Sql.Expression("your_custom_sql_here")]
public static DateTime MyAddFunction(DateTime date, TimeSpan interval) => date + interval;
```

In this case it will work on client side as well.
